src/gluonts/mx/trainer/callback.py

```diff
@@ -0,0 +1,623 @@
+
+    def on_train_start(self, max_epochs: int) -> None:
+        """
+        Hook that is called prior to training. This is the very first hook to be called.
+
+        Parameters
+        ----------
+        max_epochs
+            The maximum number of epochs that training is running. The actual number of epochs may
+            be fewer if another callback hook stops training early.
+        """
+
+    def on_network_initializing_end(
+        self, training_network: nn.HybridBlock
+    ) -> None:
+        """
+        Hook that is called prior to training, after the training network has been initialized.
+        This is the first hook where the network is passed.
+
+        Parameters
+        ----------
+        training_network
+            The network that is being trained.
+        """
+
+    def on_train_epoch_start(self, training_network: nn.HybridBlock) -> None:
+        """
+        Hook that is called prior to each training epoch.
+
+        Parameters
+        ----------
+        training_network
+            The network that is being trained.
+        """
+
+    def on_validation_epoch_start(
+        self, training_network: nn.HybridBlock
+    ) -> None:
+        """
+        Hook that is called prior to each validation epoch. This hook is never called if no
+        validation data is available during training.
+
+        Parameters
+        ----------
+        training_network
+            The network that is being trained.
+        """
+
+    def on_train_batch_end(self, training_network: nn.HybridBlock) -> None:
+        """
+        Hook that is called after each training batch.
+
+        Parameters
+        ----------
+        training_network
+            The network that is being trained.
+        """
+
+    def on_validation_batch_end(
+        self, training_network: nn.HybridBlock
+    ) -> None:
+        """
+        Hook that is called after each validation batch. This hook is never called if no validation
+        data is available during training.
+
+        Parameters
+        ----------
+        training_network
+            The network that is being trained.
+        """
+
+    def on_train_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        """
+        Hook that is called after each training epoch. This method returns a boolean whether
+        training should continue.
+
+        Parameters
+        ----------
+        epoch_no
+            The current epoch (the first epoch has `epoch_no = 0`).
+        epoch_loss
+            The loss that was recorded in the last epoch.
+        training_network
+            The network that is being trained.
+        trainer
+            The trainer which is running the training.
+
+        Returns
+        -------
+        bool
+            A boolean whether the training should continue. Defaults to `True`.
+        """
+        return True
```

Comment:
Personally, I dislike to use plain types to encode behaviour. There is no immediate connection between `True` and continuing training.

In languages like Rust you would use an enum to encode different behaviour, which we could adopt.

As an idea, we could have something like that:

```class CallbackResult(Enum):
    TRAIN_STOP = 1
```

And then return that whenever we want to strop training. `None` could mean we don't want to do anything.

Reply:
Imo, enums are not really pythonic, I have hardly seen them in large Python projects.

Personally, I would prefer throwing a specific type of exception here (either `StopIteration`, but that might be misused, or something like `StopTrainingException` or `CallbackException`). We could then catch this specific type of exception in the training loop and gracefully exit.

I think exceptions are quite nice in this context since you rarely have callbacks which stop training early (I can only think of early stopping behavior actually).

Reply:
I think just `raise StopTraining` looks good to me.

Since it bubbles automatically, we might not even need to catch it in the list-thingy.

Reply:
One downside to that is that, potentially, not all callbacks are called. If the first one throws an exception and the second one logs progress at the end of each batch/epoch, it is not called. It's not too much work to catch and re-raise in the `CallbackList` though.